documentation/binding-data/tutorial-flow-data-provider.asciidoc

```diff
@@ -53,535 +112,429 @@ grid.setItems(
 });
-----
-
-== Using a ListDataProvider for Advanced In-memory Data Handling
 
-As an alternative to assigning the items in a collection directly, you can create a `ListDataProvider` that contains the items a component should use. 
+// Hook to item count change event
+dataView.addItemCountChangeListener(e ->
+        Notification.show(" " + e.getItemCount() + " items available"));
 
-Multiple components can share a single list data provider to display the same data. You can also configure the instance to filter out some items or display items in a specific order.
+// Request the item count directly
+Span itemCountSpan = new Span("Total Item Count: " + dataView.getItemCount());
+----
 
-For components like `Grid` that can be separately configured to sort data in a specific way, sorting configured in the data provider is only used as a fallback. The fallback is used if no sorting is defined in the component, or if the order between items is considered equal by the component's sorting definition. Components update automatically when you change sorting in the data provider.
+If you are using an in-memory data set, you can also apply filters through the data view object.
+The filtered list is automatically updated to the UI.
 
-*Example*: Defining differing sort orders in the `ListDataProvider` and components.
+For example, you can use a list data view to filter items based on a property as follows:
 
 [source, java]
 ----
-ListDataProvider<Person> dataProvider =
-        DataProvider.ofCollection(persons);
+List<Person> allPersons = repo.findAll();
+GridListDataView<Person> gridDataView = grid.setItems(allPersons);
 
-dataProvider.setSortOrder(Person::getName,
-        SortDirection.ASCENDING);
-
-Grid<Person> grid = new Grid<>(Person.class);
-// The grid shows the persons sorted by name
-grid.setDataProvider(dataProvider);
-
-// Makes the combo box show persons in descending order
-button.addClickListener(event -> {
-    dataProvider.setSortOrder(Person::getName,
-            SortDirection.DESCENDING);
-});
+// Filter Persons younger 20 years
+gridDataView.setFilter(p -> p.getAge() < 20);
 ----
 
-=== Filtering In-memory Data
+== Recycling Data Binding Logic
 
-You can configure the list data provider to always apply a specific filter to limit which items display, or to filter by data that is not included in the displayed item caption. 
+In large applications, you typically have multiple places where you display the same data type in a listing component.
+You can use multiple approaches to share the lazy data binding logic.
 
-*Example*: Defining a `ListDataProvider` with a filter. 
+One way is to use a domain-object specific component implementation by extending a listing component to handle the application-specific data binding.
+This approach allows sharing also other common configuration.
 
 [source, java]
 ----
-ListDataProvider<Person> dataProvider =
-        DataProvider.ofCollection(persons);
+@SpringComponent
```

Comment:
This is extremely bad example since it declares a Vaadin component as a Spring component but doesn't assign any scope which means the default scope which is singleton and this is always wrong.
Either remove the annotation or declare meaningful scope (e.g. prototype).

Reply:
The component can be a route target then it will be instantiated with autowiring.
(and then Spring annotation can be removed)
otherwise the annotation is needed but then scope should be provided.

Reply:
Good catch! Done